decode/decode.go

```diff
@@ -96,6 +96,243 @@ func FieldsFromKayvee(line string) (map[string]interface{}, error) {
 	return m, nil
 }
 
+// MetricsRoute represents a metrics kv log route
+type MetricsRoute struct {
+	Series     string
+	Dimensions []string
+	ValueField string
+	RuleName   string
+}
+
+// AnalyticsRoute represents an analytics kv log route
+type AnalyticsRoute struct {
+	Series   string
+	RuleName string
+}
+
+// NotificationRoute represents a notification kv log route
+type NotificationRoute struct {
+	Channel  string
+	Icon     string
+	Message  string
+	User     string
+	RuleName string
+}
+
+// AlertRoute represents an alert kv log route
+type AlertRoute struct {
+	Series     string
+	Dimensions []string
+	StatType   string
+	ValueField string
+	RuleName   string
+}
+
+func getStringValue(json map[string]interface{}, key string) string {
+	val, ok := json[key]
+	if !ok {
+		return ""
+	}
+
+	str, ok := val.(string)
+	if !ok {
+		return ""
+	}
+
+	return str
+}
+
+func getStringArray(json map[string]interface{}, key string) []string {
+	val, ok := json[key]
+	if !ok {
+		return []string{}
+	}
+
+	strArray, ok := val.([]string)
+	if !ok {
+		return []string{}
+	}
+
+	return strArray
+}
+
+// LogRoutes a type alias to make it easier to add route specific filter functions
+type LogRoutes []map[string]interface{}
+
+// MetricsRoutes filters the LogRoutes and returns a list of MetricsRoutes structs
+func (l LogRoutes) MetricsRoutes() []MetricsRoute {
+	routes := []MetricsRoute{}
+
+	for _, route := range l {
+		tipe := getStringValue(route, "type")
+		if tipe != "metrics" {
+			continue
+		}
+
+		series := getStringValue(route, "series")
+		dimensions := getStringArray(route, "dimensions")
+		valueField := getStringValue(route, "value_field")
+		ruleName := getStringValue(route, "rule")
+
+		if series == "" { // TODO: log error
+			continue
+		}
+		if valueField == "" {
+			valueField = "value"
+		}
+
+		routes = append(routes, MetricsRoute{
+			Series:     series,
+			Dimensions: dimensions,
+			ValueField: valueField,
+			RuleName:   ruleName,
+		})
+	}
+
+	return routes
+}
+
+// AnalyticsRoutes filters the LogRoutes and returns a list of AnalyticsRoutes structs
+func (l LogRoutes) AnalyticsRoutes() []AnalyticsRoute {
+	routes := []AnalyticsRoute{}
+
+	for _, route := range l {
+		tipe := getStringValue(route, "type")
+		if tipe != "analytics" {
+			continue
+		}
+
+		series := getStringValue(route, "series")
+		ruleName := getStringValue(route, "rule")
+
+		if series == "" { // TODO: log error
+			continue
+		}
```

Comment:
That's an interesting idea, but I'm a little afraid it'll hurt the usability of the interface.  Instead of consumers doing something like this:

```go
kvmeta = decode.ExtractKVMeta(fields)
routes = kvmeta.AnalyticsRoutes()
```

They'd have have to something like this:

```go
kvmeta = decode.ExtractKVMeta(fields)
routes := []AnalyticsRoute{}
for _, route := kvmeta.Routes {
	// Optionally, add a `route.IsAnalytics()` method

	analytics, err := route.ToAnalyticsRoute()
	if err != nil { // TODO: log unexpected errors
		continue
	}

	routes = append(routes, analytics)
}
```